relay-server/src/http.rs

```diff
@@ -0,0 +1,299 @@
+}
+
+pub enum RequestBuilder {
+    Actix(ClientRequestBuilder),
+    Reqwest {
+        builder: reqwest::RequestBuilder,
+
+        /// The content encoding that this builder object implements on top of reqwest, which does
+        /// not support request encoding at all.
+        http_encoding: HttpEncoding,
+    },
+}
+
+impl RequestBuilder {
+    pub fn reqwest(builder: reqwest::RequestBuilder) -> Self {
+        RequestBuilder::Reqwest {
+            builder,
+
+            // very few endpoints can actually deal with request body content-encoding. Outside of
+            // store/envelope this is almost always identity because there's no way to do content
+            // negotiation on request.
+            http_encoding: HttpEncoding::Identity,
+        }
+    }
+
+    pub fn actix(builder: ClientRequestBuilder) -> Self {
+        RequestBuilder::Actix(builder)
+    }
+
+    pub fn no_default_headers(&mut self) -> &mut Self {
+        match self {
+            RequestBuilder::Actix(builder) => {
+                builder.no_default_headers();
+            }
+            RequestBuilder::Reqwest { .. } => {
+                // can only be set on client
+            }
+        }
+
+        self
+    }
+
+    pub fn finish<E>(self) -> Result<Request, E>
+    where
+        E: From<reqwest::Error> + From<ActixError>,
+    {
+        match self {
+            RequestBuilder::Actix(mut builder) => Ok(Request::Actix(builder.finish()?)),
+            RequestBuilder::Reqwest {
+                builder,
+                http_encoding: _,
+            } => Ok(Request::Reqwest(builder.build()?)),
+        }
+    }
+
+    /// Add a new header, not replacing existing ones.
+    pub fn header(&mut self, key: &str, value: &[u8]) -> &mut Self {
+        match self {
+            RequestBuilder::Actix(builder) => {
+                builder.header(key, value);
+            }
+            RequestBuilder::Reqwest { builder, .. } => {
+                take_mut::take(builder, |b| b.header(key, value))
+            }
+        }
+
+        self
+    }
+
+    pub fn body<E>(self, body: Binary) -> Result<Request, E>
+    where
+        E: From<reqwest::Error> + From<ActixError> + From<io::Error>,
+    {
+        match self {
+            RequestBuilder::Actix(mut builder) => Ok(Request::Actix(builder.body(body)?)),
+            RequestBuilder::Reqwest {
+                mut builder,
+                http_encoding,
+            } => {
+                let body = match http_encoding {
+                    HttpEncoding::Identity => {
+                        builder = builder.header("Content-Encoding", "identity");
+                        body.as_ref().to_vec()
+                    }
+                    HttpEncoding::Deflate => {
+                        builder = builder.header("Content-Encoding", "deflate");
+                        let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
+                        encoder.write_all(body.as_ref())?;
```

Comment:
It would be great to do this in a streaming fashion. IIRC, actix web client does that internally.

Reply:
The reqwest body can only be constructed from various in-memory strings and vectors, and async streams.

I can try async stream. Do you think this needs to be on a separate threadpool or should we just do `.read()` on poll?

Reply:
Have a look how actix internally abstracts this. In all cases, I would propose to poll directly and not in a separate thread. Orchestration and resources should be managed on the service layer, not in here.

Reply:
Another point on threads: Eventually the upstream should become multithreaded, but that's much easier once actix client is gone.